extensions/src/AWSSDK.Extensions.CborProtocol/CborWriterExtensions.cs

```diff
@@ -0,0 +1,122 @@
+                    // WriteInt64 serializes it into the smallest CBOR type representation
+                    // that can contain its value without loss of precision.
+                    writer.WriteInt64((long)value);
+                    return;
+                }
+
+                if (value >= 0 && value <= ulong.MaxValue)
+                {
+                    // If the value is non-negative and fits within the unsigned 64-bit range,
+                    // WriteUInt64 serializes it into the smallest possible CBOR type representation.
+                    writer.WriteUInt64((ulong)value);
+                    return;
+                }
+            }
+
+            // Check if value can safely be represented as float32
+            float floatCandidate = (float)value;
+            if ((double)floatCandidate == value)
+            {
+                WriteOptimizedNumber(writer, floatCandidate);
+                return;
+            }
+
+            // If none of the above conditions are satisfied, write the value as a double.
+            writer.WriteDouble(value);
+        }
+
+        /// <summary>
+        /// Writes a float using the smallest CBOR representation that preserves value and precision.
+        /// This method uses manual encoding to avoid writing as a half-precision float.
+        /// </summary>
+        /// <param name="writer">The CBOR writer to use.</param>
+        /// <param name="value">The float value to write.</param>
+        public static void WriteOptimizedNumber(this CborWriter writer, float value)
+        {
+            // If the value is an integer (without fractional part), write it as Int64 or UInt64.
+            if (value % 1 == 0)
+            {
+                if (value >= long.MinValue && value <= long.MaxValue)
+                {
+                    // If the value fits within the signed 64-bit integer (long) range,
+                    // WriteInt64 serializes it into the smallest CBOR type representation
+                    // that can contain its value without loss of precision.
+                    writer.WriteInt64((long)value);
+                    return;
+                }
+
+                if (value >= 0 && value <= ulong.MaxValue)
+                {
+                    // If the value is non-negative and fits within the unsigned 64-bit range,
+                    // WriteUInt64 serializes it into the smallest possible CBOR type representation.
+                    writer.WriteUInt64((ulong)value);
+                    return;
+                }
+            }
+
+            // Manual encoding to avoid half-precision floats
+            var bytes = new byte[5];
+            bytes[0] = 0xFA; // CBOR float32 marker
+            BitConverter.GetBytes(value).CopyTo(bytes, 1);
+
+            // Ensure the bytes are in the correct endian order for CBOR.
+            if (BitConverter.IsLittleEndian)
+                Array.Reverse(bytes, 1, 4);
```

Comment:
```suggestion
            Span<byte> bytes = new byte[5];
            bytes[0] = 0xFA; // CBOR float32 marker
            BinaryPrimitives.WriteSingleBigEndian(bytes.AsSpan(1), value);
```

(and add a `using System.Buffers.Binary;` at the top)

Reply:
Unfortunately, we can't use this approach because the SDK targets multiple .net versions that don't support it.
Currently with V4 we support `net8`, `netstandard2.0`, `netcore3.1` and `netframework4.7.2`, I believe `net8` is the only target that fully support this approach.